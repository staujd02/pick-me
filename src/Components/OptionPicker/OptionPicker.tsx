import React from 'react';

export type OptionPickerProps = {
  finishedPickingOptionsCallback: (chosenOptions: Option[]) => void
};
export type Option = {
  image: string
  title: string
}

const OptionPicker: React.FC<OptionPickerProps> = (props: OptionPickerProps) => {
  return (
    <div className="option-picker"></div>
  );
}

export default OptionPicker;
