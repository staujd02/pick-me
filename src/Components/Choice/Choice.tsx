import React from 'react';
import { Option } from '../OptionPicker/OptionPicker';

export type ChoiceProps = {
    optionsToChooseFrom: Option[]
    finishedChoosingOption: (chosenOption: Option) => void 
}
export type ChoiceState = {}

class Choice extends React.Component<ChoiceProps, ChoiceState>{
    render(){
        return (
            <div></div>
        );
    }
}

export default Choice;