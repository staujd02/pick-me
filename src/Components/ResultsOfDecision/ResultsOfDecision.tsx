import React from 'react';
import { Option } from '../OptionPicker/OptionPicker';

export type ResultsOfDecisionProps = {
    chosenOption: Option
};
export type ResultsOfDecisionState = {};

class ResultsOfDecision extends React.Component<ResultsOfDecisionProps, ResultsOfDecisionState> {
    render(){
        return (
            <div></div>
        );
    }
}

export default ResultsOfDecision;