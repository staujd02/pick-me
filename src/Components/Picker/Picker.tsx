import React from 'react';
import OptionPicker, { Option } from '../OptionPicker/OptionPicker';
import Choice from '../Choice/Choice';
import ResultsOfDecision from '../ResultsOfDecision/ResultsOfDecision';

export type PickerProps = {}
export type PickerState = {
    isPicking: boolean
    isChoosing: boolean
    hasChosen: boolean
    availableOptions: Option[],
    chosenOption: Option
}

class Picker extends React.Component<PickerProps, PickerState> {
    constructor(props: PickerProps) {
        super(props);
        this.state = {
            isPicking: false,
            isChoosing: false,
            hasChosen: false,
            availableOptions: [],
            chosenOption: {
                image: '',
                title: ''
            }
        }
    }

    onClick = () => {
        this.setState({
            isPicking: true
        });
    }

    optionsPicked = (optionsChosen: Option[]) => {
        this.setState({
            isChoosing: true,
            availableOptions: optionsChosen
        });
    }

    optionChosen = (option: Option) => {
        this.setState({
            hasChosen: true,
            chosenOption: option
        });
    }

    render() {
        const { isPicking, isChoosing, hasChosen, availableOptions, chosenOption } = this.state;
        return (
            <React.Fragment>
                {!isPicking && <button onClick={this.onClick} id="setup-choice">Make Choice</button>}
                {isPicking && !isChoosing && <OptionPicker finishedPickingOptionsCallback={this.optionsPicked} />}
                {isChoosing && !hasChosen && <Choice optionsToChooseFrom={availableOptions} finishedChoosingOption={this.optionChosen}/>}
                {hasChosen && <ResultsOfDecision chosenOption={chosenOption} />}
            </React.Fragment>
        );
    }

}

export default Picker;