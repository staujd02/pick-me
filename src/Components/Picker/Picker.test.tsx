import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { formatHTML } from '../../TestUtilities/htmlFormatter';
import Picker from './Picker';
import OptionPicker from '../OptionPicker/OptionPicker';
import Choice from '../Choice/Choice';
import ResultsOfDecision from '../ResultsOfDecision/ResultsOfDecision';

describe('The Picker', () => {

    let wrapper: ReactWrapper<PickerProps, PickerState, Picker>;

    beforeEach(() => {
        wrapper = mount(<Picker />);
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());

    it('does not ask the user to set up an option by default', () => {
        let c = wrapper.instance().optionsPicked;
        expect(wrapper.contains(<OptionPicker finishedPickingOptionsCallback={c} />)).toEqual(false);
    });

    it('does not ask the user to make a choice by default', () => {
        let opts: Option[] = [];
        let c = wrapper.instance().optionChosen;
        expect(wrapper.contains(<Choice optionsToChooseFrom={opts} finishedChoosingOption={c} />)).toEqual(false);
    });

    it('does not show the user a choice has been made by default', () => {
        let opt: Option = { image: "", title: "" };
        expect(wrapper.contains(<ResultsOfDecision chosenOption={opt} />)).toEqual(false);
    });

    describe('when the user chooses to set up a choice', () => {
        beforeEach(() => {
            clickSetupChoices();
            wrapper.update();
        });

        it('no longer shows the option to set up a choice', () => {
            expect(wrapper.find('#setup-choice').length).toEqual(0);
        });

        it('should show a picker for those options', () => {
            let c = wrapper.instance().optionsPicked;
            expect(wrapper.contains(<OptionPicker finishedPickingOptionsCallback={c} />)).toEqual(true);
        });
    });

    describe('given the user has selected some options', () => {

        beforeEach(() => {
            clickSetupChoices();
            wrapper.update();
        });

        it('no longer shows the option to set up a choice', () => {
            expect(wrapper.find('#setup-choice').length).toEqual(0);
        });

        describe('and is now ready to make a choice', () => {

            beforeEach(() => {
                wrapper.instance().optionsPicked([]);
                wrapper.update()
            });

            it('does not show the option to pick different choices', () => {
                expect(wrapper.find('OptionPicker').length).toEqual(0);
            });

            describe('and is now ready to see the choice', () => {

                beforeEach(() => {
                    let option: Option = {
                        image: "base64",
                        title: "I am chosen"
                    }
                    wrapper.instance().optionChosen(option);
                    wrapper.update();
                });

                it('does not show the option to make a different decision', () => {
                    expect(wrapper.find('Choice').length).toEqual(0);
                });
            });
        });
    });

    describe('when the user has made their choice', () => {

        const option: Option = {
            image: "base64",
            title: "Chosen One"
        };

        beforeEach(() => {
            wrapper.instance().optionChosen(option);
            wrapper.update();
        });

        it('displays some results', () => {
            expect(wrapper.find('ResultsOfDecision').length).toEqual(1);
        });

        it('shows the option the user picked', () => {
            expect(wrapper.find('ResultsOfDecision').prop('chosenOption')).toEqual(option);
        });
    });

    describe('when the user has picked their options', () => {

        const options = [{
            image: "base64/string",
            title: "House"
        }, {
            image: "base64/string/other",
            title: "Or A Home"
        }];

        beforeEach(() => {
            wrapper.instance().optionsPicked(options);
            wrapper.update();
        });

        it('shows the menu for the next user to pick with', () => {
            let opts = wrapper.state().availableOptions;
            let c = wrapper.instance().optionChosen;
            expect(wrapper.contains(<Choice optionsToChooseFrom={opts} finishedChoosingOption={c} />)).toEqual(true);
        });

        it('gives the user the available options', () => {
            expect(wrapper.find('Choice').prop('optionsToChooseFrom')).toEqual(options);
        });
    });

    function clickSetupChoices() {
        wrapper.find('#setup-choice').simulate('click');
    }
});
