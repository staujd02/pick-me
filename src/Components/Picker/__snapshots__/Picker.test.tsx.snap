// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Picker renders correctly 1`] = `
"
<button id=\\"setup-choice\\">Make Choice</button>
"
`;
